File: applications/node/replacement.mjs
import {readFile,writeFile,mkdir,lstat} from 'node:fs/promises';
import path from 'node:path';
import {VERSION,PROFILE,SCHEMA_ROOT,jsonFile,strictJSON,validateSchema,writeJSON,sha256,fail,deepEqual} from './base.mjs';
import {loadPackage,preserve} from './package.mjs';

// Deliberate binding revision only. This does not translate tools, graph or units.
export async function replaceBinding(source,{binding,replacementFile,revision,outDirectory}) {
  const pkg=await loadPackage(source);const prior=pkg.requirements.bindings[binding];
  if(!prior)fail('EPX_BINDING_INVALID','invalid','BIND-006','Replacement references an undeclared logical binding.',{dependency:binding});
  if(typeof revision!=='string'||!revision.trim()||revision===pkg.requirements.process.revision)fail('EPX_REPLACEMENT_REVISION_INVALID','invalid','BIND-006','Explicit replacement requires a fresh process revision.');
  const replacement=await jsonFile(replacementFile);const schema=await jsonFile(path.join(SCHEMA_ROOT,'requirements.schema.json'));
  validateSchema(replacement,{$schema:schema.$schema,$ref:'#/$defs/binding',$defs:schema.$defs},'BIND-006');
  if(deepEqual(replacement,prior))fail('EPX_REPLACEMENT_UNCHANGED','invalid','BIND-006','The supplied binding is unchanged; configure a receiving connection without editing the process.');
  const out=path.resolve(outDirectory);
  const previousBytes=pkg.payload['requirements.json'].bytes;const previousSha=sha256(previousBytes);
  const previousPath=`provenance/requirements-${previousSha}.json`;
  const reportPath=`provenance/replacement-${previousSha}-${sha256(revision).slice(0,16)}.json`;
  if(pkg.files.includes(previousPath)||pkg.files.includes(reportPath))fail('EPX_REPLACEMENT_PROVENANCE_CONFLICT','invalid','BIND-006','The intended provenance path already exists; preserve it and select another output revision.');
  const requirements=structuredClone(pkg.requirements);requirements.process.revision=revision;requirements.bindings[binding]=replacement;
  const impactedNodes=Object.entries(requirements.operations).filter(([,op])=>op.binding===binding).map(([node])=>node);
  const direct=new Set(impactedNodes);const affected=new Set(direct);const main=pkg.main;
  // Impact analysis only. CWL still owns dataflow execution and connections remain unchanged.
  for(let changed=true;changed;) {
    changed=false;
    for(const [name,step] of Object.entries(main.steps)) {
      const node=`main/${name}`;if(affected.has(node))continue;
      const sources=Object.values(step.in??{}).flatMap(input=>{const raw=typeof input==='string'?input:input.source;return raw===undefined?[]:Array.isArray(raw)?raw:[raw];});
      if(sources.some(source=>typeof source==='string'&&source.includes('/')&&affected.has(`main/${source.split('/')[0]}`))){affected.add(node);changed=true;}
    }
  }
  const change={format:'epx-binding-replacement',version:VERSION,profile:PROFILE,processId:requirements.process.id,previousRevision:pkg.requirements.process.revision,revision,binding,previousPackageMetadataSha256:pkg.metadataSha256,previousRequirements:{path:previousPath,sha256:previousSha},previousBinding:prior,replacementBinding:replacement,directlyImpactedNodes:impactedNodes,impactedNodes:[...affected].sort(),impactedAcceptance:requirements.acceptance.filter(c=>affected.has(c.node)).map(c=>c.id),impactedResources:Object.entries(requirements.resources).filter(([,r])=>r.binding===binding).map(([id])=>id),toolContractsChanged:false,graphChanged:false,resourceMappingsChanged:false,providerEquivalenceClaimed:false,previousResultsReusable:false,humanApprovalsReusable:false,execution:'not-run',nextAction:'Validate receiving identity/resources and execute new acceptance evidence; broader tool/resource mapping changes require explicit authoring.'};
  await preserve(pkg.root,out);await mkdir(path.join(out,'provenance'),{recursive:true});await writeFile(path.join(out,previousPath),previousBytes);await writeJSON(path.join(out,reportPath),change);await writeJSON(path.join(out,'requirements.json'),requirements);
  const metadata=structuredClone(pkg.metadata);const root=metadata['@graph'].find(entity=>entity['@id']==='./');
  const parts=Array.isArray(root.hasPart)?root.hasPart:[root.hasPart];root.hasPart=parts;
  for(const filename of ['requirements.json',previousPath,reportPath]) {
    const bytes=await readFile(path.join(out,filename));let entity=metadata['@graph'].find(entity=>entity['@id']===filename);
    if(!entity){entity={'@id':filename,'@type':'File'};metadata['@graph'].push(entity);root.hasPart.push({'@id':filename});}
    entity.sha256=sha256(bytes);entity.contentSize=String(bytes.length);
  }
  await writeJSON(path.join(out,'ro-crate-metadata.json'),metadata);
  const validated=await loadPackage(out);
  for(const filename of pkg.files.filter(filename=>!['requirements.json','ro-crate-metadata.json'].includes(filename)))if(!Buffer.from(await readFile(path.join(out,filename))).equals(await readFile(path.join(pkg.root,filename))))fail('EPX_REPLACEMENT_PRESERVATION_FAILED','execution-failed','BIND-006','Replacement changed an unaffected artifact.',{dependency:filename});
  return {format:'epx-replacement-result',version:VERSION,status:'valid',process:validated.requirements.process,packageMetadataSha256:validated.metadataSha256,change,changeReport:reportPath,engineeringCalls:0};
}

File: src/epx/replacement.py
"""Explicit, limited provider replacement; never an automatic fallback."""

from copy import deepcopy
from pathlib import Path

from . import VERSION, jsonio
from .errors import invalid
from .package import load, open_root, preserve


def replace(source, binding, replacement, revision, output):
    """Retain graph/contracts, revise one intended binding, expose review impact.

    This operation does not prove another application's engineering equivalence.
    The receiving environment must supply a new local binding and preflight it.
    """
    destination = Path(output).resolve()
    with open_root(source) as root:
        package = load(root)
        previous = package.requirements
        if binding not in previous["bindings"]:
            raise invalid("REPLACEMENT_BINDING", "BIND-006", "The named binding does not exist", dependency=binding)
        if not isinstance(revision, str) or not revision.strip() or revision == previous["process"]["revision"]:
            raise invalid("REPLACEMENT_REVISION", "BIND-006", "Explicit replacement requires a distinct nonempty process revision")
        updated = deepcopy(previous)
        updated["bindings"][binding] = deepcopy(replacement)
        updated["process"]["revision"] = revision
        jsonio.validate(updated, jsonio.load_schema("requirements"), requirement="BIND-006")
        if jsonio.equal(previous["bindings"][binding], replacement):
            raise invalid("REPLACEMENT_UNCHANGED", "BIND-006", "Replacement must change the intended binding; local connection changes do not revise a process")
        direct = {node for node, operation in previous["operations"].items() if operation["binding"] == binding}
        impacted = set(direct)
        changed = True
        while changed:
            changed = False
            for step_name, step in package.main["steps"].items():
                node = "main/" + step_name
                for parameter in step["in"].values():
                    sources = parameter.get("source", []) if isinstance(parameter, dict) else parameter
                    if isinstance(sources, str):
                        sources = [sources]
                    if node not in impacted and any("/" in item and "main/" + item.split("/")[0] in impacted for item in sources or []):
                        impacted.add(node)
                        changed = True
        original_bytes = (root / "requirements.json").read_bytes()
        original_digest = jsonio.digest(original_bytes)
        provenance_path = "provenance/requirements-" + original_digest + ".json"
        edit_path = "provenance/replacement-" + original_digest + ".json"
        if (root / edit_path).exists():
            raise invalid("REPLACEMENT_PROVENANCE", "BIND-006", "Replacement provenance target already exists")
        report = {
            "format": "epx-replacement", "version": VERSION,
            "processId": previous["process"]["id"], "previousRevision": previous["process"]["revision"],
            "revision": revision, "binding": binding,
            "previousRequirementsSha256": original_digest, "previousRequirements": provenance_path,
            "before": previous["bindings"][binding], "after": replacement,
            "directNodes": sorted(direct), "impactedNodes": sorted(impacted),
            "acceptanceRequiresRerun": [item["id"] for item in previous["acceptance"] if item["node"] in impacted],
            "approval": "not-requested", "historicalEvidence": "retained; never carried forward as new acceptance",
            "reviewRequired": ["tool input/output contracts", "resource identities and revisions", "data mappings and downstream acceptance"],
            "engineeringEquivalence": "not-established", "engineeringCalls": 0,
        }
        preserve(root, destination)
        (destination / provenance_path).parent.mkdir(parents=True, exist_ok=True)
        (destination / provenance_path).write_bytes(original_bytes)
        jsonio.write(destination / edit_path, report)
        jsonio.write(destination / "requirements.json", updated)
        metadata = deepcopy(package.metadata)
        dataset = next(item for item in metadata["@graph"] if item["@id"] == "./")
        for relative in ("requirements.json", provenance_path, edit_path):
            payload = (destination / relative).read_bytes()
            entity = next((item for item in metadata["@graph"] if item["@id"] == relative), None)
            if entity is None:
                entity = {"@id": relative, "@type": "File"}
                metadata["@graph"].append(entity)
                dataset["hasPart"].append({"@id": relative})
            entity.update(sha256=jsonio.digest(payload), contentSize=str(len(payload)))
        jsonio.write(destination / "ro-crate-metadata.json", metadata)
        load(destination)
        return {"status": "replaced", "destination": str(destination), **report}
